### PR TITLE
Don't leak Workspace objects from AnalyzerAndOptions

### LINKED ISSUE #2777: we are leaking Solution in AnalyzerManager
we are leaking Solution and PreviewWorkspace through this

http://source.roslyn.io/#Microsoft.CodeAnalysis/AnalyzerManager.cs,8b48a4a85a533bd8,references

basically, AnalyzerAndOption contains WorkspaceAnalyzerOption that hold onto Workspace. and once put it there, it never goes away.

### PR DESCRIPTION
**Pull request by JohnHamby**
Prevent AnalyzerAndOptions objects from holding on to Workspace objects from that can be embedded in WorkspaceAnalyzerOptions objects.

Fixes #2777.

@heejaechang @mavasani @srivatsn might be interested.

### CODE REVIEW

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.AnalyzerAndOptions.cs [hunk 1]
```diff
@@ -18,7 +18,13 @@ public AnalyzerAndOptions(DiagnosticAnalyzer analyzer, AnalyzerOptions analyzerO
                 Debug.Assert(analyzerOptions != null);
 
                 Analyzer = analyzer;
-                _analyzerOptions = analyzerOptions;
+                // _analyzerOptions is used only as a key for equality testing, and only the properties
```
Comment (heejaechang) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.AnalyzerAndOptions.cs]: @mavasani will this break analyzer that depends on doing var options = (WorkspaceAnalyzerOptions)context.analyzeroption?

...

this still leave the leak where AnalyzerAndOptions added to the map never gets removed.

**Comment by heejaechang** [comment]
@JohnHamby I think this removes some leaks but still left leak on the map that hold onto AnalyzerAndOptions. and the values of the map.

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.AnalyzerAndOptions.cs (hunk 1 above)
Comment (JohnHamby) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.AnalyzerAndOptions.cs]: Well, it makes an integration test fail, so it breaks something.

Comment (heejaechang) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.AnalyzerAndOptions.cs]: I think there are some VS only analyzer that depends on the implementation detail. what I mean is that it knows it is WorkspaceAnalyzerOption type and then try to cast to that type.

Comment (mavasani) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.AnalyzerAndOptions.cs]: @heejaechang No, this change won't break those analyzers. AnalyzerOptions created here by John here is just used for the key of a map, not actually passed to any analyzer. They will get original instance.

@JohnHamby  which test failed? It should be unrelated to this change.

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.AnalyzerAndOptions.cs [hunk 2]
```diff
@@ -18,7 +18,13 @@ public AnalyzerAndOptions(DiagnosticAnalyzer analyzer, AnalyzerOptions analyzerO
...
+                // declared by AnalyzerOptions are appropriate for use in the comparison. (For example,
+                // two AnalyzersAndOptions that differ only by Workspace ought to be treated as equal.)
+                // If the supplied analyzerOptions is an instance of a derived type, holding on to that object
+                // can introduce a memory leak, as well as causing false negatives in comparison. Avoid these
+                // issues by creating a new object with only the desired properties. 
+                _analyzerOptions = new AnalyzerOptions(analyzerOptions.AdditionalFiles);
```
Comment (mavasani) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.AnalyzerAndOptions.cs]: John, you may also want to add a note that this a temporary fix. Once #2706 is fixed, we can get rid of `WorkspaceAnalyzerOptions` hack and seal `AnalyzerOptions`.

**Comment by mavasani** [comment]
:+1:

**Comment by mavasani** [comment]
@heejaechang Which leaks, can you expand? The map here: http://source.roslyn.io/Microsoft.CodeAnalysis/R/8b48a4a85a533bd8.html is now keyed on instance of analyzer + the instance of AnalyzerOptions. For project specific analyzers, they explicitly clear all analyzer state on AnalyzerManager. For host specific analyzers, they live for entire VS session, so the entries for them also lives entire VS session.

**Comment by heejaechang** [comment]
@mavasani @JohnHamby I dont understand. the leak was AnalyerAndOption that point to PreviewWorkspace left alive even after Preview window is closed.

so, this change cut link from AnalyzerAndOption to PreviewWindow. but the AnalyzerAndOption is still in the map. if AnalyzerAndOption was for PreviewWindow, then that should removed as well.

**Comment by mavasani** [comment]
@heejaechang AnalyzerAndOption is still the key, but now it doesn't hold onto the AnalyzerOptions instance which has workspace, it has a new AnalyzerOptions instance which is guaranteed to not store the Workspace. So how will we leak workspace here?

**Comment by mavasani** [comment]
Note that AnalyzerAndOptions just hold onto instance of analyzer and core analyzer options, i.e. additional file streams, which are not related to a workspace.

**Comment by mavasani** [comment]
Let me clarify futher, what we want is that we execute RegisterCompilationStartAction exactly once for a given tuple of {analyzer, analyzerOptions, compilation}. If there are multiple workspace that share these things, then nested registered actions will be shared amongst workspace, and even when the preview workspace goes, we have to retain this entry for subsequent queries made from VS Workspace.

**Comment by heejaechang** [comment]
@mavasani @JohnHamby are you saying the leak is by design? I am not sure why this became argument?. I saw there was 4-5 (can't remember exact number, but it was not 1) of them in the dump. so, it definitely hold onto multiple different previewworksace. so you are saying we must since there is a chance same thing can get used?

**Comment by heejaechang** [comment]
@JohnHamby @mavasani even for host analyzers, so if I change AnalyzerOptions, then, it leaks every time?

**Comment by mavasani** [comment]
@heejaechang I am not sure I understand how you are defining a leak here. Prior to John's fix, the entry generated in compilationScope map could hold onto preview Workspace for lifetime of analyzer, which for host analyzers is entire VS session. John's change fixes the leak. However, even when preview workspace is disposed, the host analyzer still lives on for entire VS session. If a new preview workspace is created, and it asks for host analyzer's actions for the same compilation as previously disposed compilation had asked, we should not re-execute RegisterCompilationStartAction for the same analyzer instance for the same compilation. The entry created in compilationScopeMap must live for the lifetime of analyzer, not the lifetime of workspace which happened to first ask for its entire actions, that is not a leak but a necessity for the analyzer execution semantics that we have defined.

**Comment by mavasani** [comment]
Regarding your second question, when a new AnalyzerOptions is created (say you change additional file for a project), we will end up holding onto the entry for entire lifetime of analyzer as we don't know if in future the analyzer host will ask us about analyzer actions with older options. Unless the host knows that it will never requery with the older AnalyzerOptions instance, we can't clear that entry. However, as user types and compilations get collected, the entries in the conditional weak table stored as value on compilationScopeMap will get collected and hence we only hold onto the AnalyzerOptions instance for lifetime of analyzer. If you strongly feel about fixing leak of analyzer options (which requires user to edit additional files, which are uncommon) then we can probably  further break down compilationScopeMap into ConcurrentDictionary<DiagnosticAnalyzer, ConditionalWeakTable<AnalyzerOptions, ConditionalWeakTable<Compilation, Task<HostCompilationStartAnalysisScope>>>>.

**Comment by mavasani** [comment]
@heejaechang @JohnHamby how about this: instead of storing the instance of AnalyzerOptions in AnalyzerAndOptions, can we instead store `new AnalyzerOptions(analyzerOptions.AdditionalFiles).GetHashCode()`? Its highly unlikely that there will be a collision, and even does so in a rare case, we will just execute RegisterCompilationStartAction twice for same compilation. It will ensure no leak of instances of AnalyzerOptions though.

**Comment by JohnHamby** [comment]
I can no longer repro the integration test failure, so I suspect it was an artifact of another issue.

If I've understood the discussion above, and I'm not sure that I do, this change prevents leaking Workspace instances (which is what the original issue is about).

Leaking instances of AnalyzerOptions doesn't seem terribly significant, but maybe I'm not understanding. I'm not keen on Manish's suggestion to use a hash code as a key, because it makes the system nondeterministic.

Are we better off with this change or without it? If we're better off with it, could I get another signoff?

@heejaechang @srivatsn @pharring

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.AnalyzerAndOptions.cs (hunk 1 above)
Comment (JohnHamby) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.AnalyzerAndOptions.cs]: After rebooting, I no longer see an integration test failure.

**Comment by mavasani** [comment]
I think the more important thing is not leaking workspaces. Changes to additional files (which can leak AnalyzerOptions) are not going to be that harmful, and we are going to fix that one very soon after 1.0 when we add support for Workspace analyzers (which is also very important). I am :+1: for current PR.

**Comment by heejaechang** [comment]
:+1:

**Pull request merged**